Mine Page: Adding Charts Display of the Expenses

LiorKGOW opened the pull request:
Adding the Pie and the Bar Charts to the Mine Page, which display the House expenses in another way.
The expenses are presented in the charts in the total amount, sorted by month.
Added a filter to choose which year of expenses are presented in the charts.

Addressing Issue: #102 

Screenshots of the Charts from the Mine Page:

Charts View of the Expenses:

![Mine Page - Charts View of the Expenses](https://user-images.githubusercontent.com/93318917/169768104-2f58bf48-3b16-46a4-b6bd-00304c106130.png)

New Expense shows in the charts after being added:

![Mine Page - Charts View of the Expenses - new expense added to the charts](https://user-images.githubusercontent.com/93318917/169768134-d4600999-14fc-4f38-a2ba-ce2a708e0125.png)

ShellyGolden commented:
Hi
Can you add screenshots? 
Thanks 🙂

LiorKGOW commented:
> Hi Can you add screenshots? Thanks 🙂

Sure thing 👍

ronyyosef commented:
@LiorKGOW Why did you get out of Shelly's branch? I don't think it really needs to be dependent on her PR

ronyyosef commented:
Looks pretty good
What do you think about adding a filter of months so the user can see the chart with information about a specific month?

shlior7 commented:
> Looks pretty good What do you think about adding a filter of months so the user can see the chart with information about a specific month?

I think showing the charts according to months should be the default and there is no need for any added filtering or filtering logic. maybe show different data on each chart would be a good idea wyt?

shlior7 commented on templates/house_view/mine_chart.html:
```diff
@@ -0,0 +1,29 @@
+{% load static %}
+
+<script type="text/javascript" src="{% static 'js/random_colors.js' %}" ></script>
+<script>
+    $(document).ready(function () {
+        window.randomColors = [{%for amount in amounts%} randomColor(),{%endfor%}];
+        window.pie_chart = "right_chart"
+        window.bar_chart = "left_chart"
+    });
+</script>
+
+{{ categories|json_script:"categories" }}
+{{ amounts|json_script:"amounts" }}
+
+<script src="{%static 'js/pie_chart.js'%}" type="text/javascript"></script>   
+<script src="{%static 'js/bar_chart.js'%}" type="text/javascript"></script>   
+
+<body>
+  <div>
+    <div class="chart_card">
+      <div class="chart_box">
+        <canvas id="right_chart"></canvas>
+      </div>
+      <div class="chart_box">
+        <canvas id="left_chart"></canvas>
+      </div>
+    </div>
+  </div>
+</body>
```
Nicely done! :)

shlior7 replied:
please take a look at my new PR https://github.com/redhat-beyond/Xpense/pull/132
this refactored the way of use of the charts and made them more customizable, this way you can change the title or any things you like in the charts by adding more params to the js functions.
please consider cherry pick my branch and add those changes to this PR and modify your use of the charts in the mine page.

LiorKGOW replied:
Nice idea @shlior7 I'll use it here 👍

LiorKGOW commented:
> @LiorKGOW Why did you get out of Shelly's branch? I don't think it really needs to be dependent on her PR

I thught it could ease the need to rebase, because we were working on the same files

LiorKGOW commented:
> Looks pretty good What do you think about adding a filter of months so the user can see the chart with information about a specific month?

I have changed the way they are presented.
The charts show the expenses by months, and you could filter, with a choice by which year is represented

shlior7 commented on static/css/mine.css:
```diff
@@ -0,0 +1,55 @@
+  bottom: 30%;
+  right: 0;
+  z-index: 9999;
+}
```
empty line in EOF(end of file)

LiorKGOW commented:
rebased and repushed after #128 #129 #132 were merged to main 👍

adiabramovitch commented on house/views.py:
```diff
@@ -81,15 +120,17 @@ def house_create(request):
-                               city=cleaned_data['city'],
-                               children=cleaned_data['children'],
-                               income=cleaned_data['income'])
+            House.create_house(
+                user=request.user,
+                name=cleaned_data['name'],
+                public=cleaned_data['public'],
+                parent_profession_1=cleaned_data['parent_profession_1'],
+                parent_profession_2=cleaned_data['parent_profession_2'],
+                country=cleaned_data['country'],
+                city=cleaned_data['city'],
+                children=cleaned_data['children'],
+                income=cleaned_data['income'],
+            )
```
it seems like this is from your commit and it is just formatting, please remove and add to a different PR.

LiorKGOW replied:
Fixed 👍

adiabramovitch commented on templates/global/chart.html:
```diff
@@ -13,7 +13,6 @@
     });
 </script>
 
-
 <body>
```
you can add this to the formatting PR.

Yoavsc129 commented on tests/my_house_tests.py:
```diff
@@ -0,0 +1,128 @@
+        response = client.post('/house/add_expense/', EXPENSE_BAD_FORM_DATA_AMOUNT)
+        assert len(Expenses.objects.all()) == 0
+        assert response.status_code == 200
+
```
Like with Shelly's The EXPENSE_BAD_FORM_DATA_ are mismatched between the two tests

LiorKGOW replied:
Fixed 👍

Yarboa reviewed (commented):
Nice work

The tests here covering model tests,
You had obviously updates the html templates,
DYT, you should update client tests?
related to this?
include 'house_view/mine_chart.html'
include 'house_view/expenses_table.html'

shlior7 reviewed (approved):
LGTM

ronyyosef commented on house/forms.py:
```diff
@@ -38,3 +39,13 @@ class ExpenseForm(forms.ModelForm):
         fields = ('date', 'amount', 'category', 'description')
+
+
+class YearFilterForm(forms.Form):
```
Did you implement the logic of this form

LiorKGOW replied:
Yes, this form filter the years from 2020 to the current year
I have changed it a bit, maybe now it would be more understandable
What do you think? @ronyyosef

shlior7 replied:
I think this method could be in inline but thats just my own preference

LiorKGOW replied:
I think it's more understandable now 😇

ronyyosef replied:
Looks great!

ronyyosef commented on house/views.py:
```diff
@@ -47,7 +49,39 @@ def house_view(request):
+        expenses_list = expenses_list.filter(date__year=datetime.now().year)
+
+    expenses_sum_by_month = expenses_list.values('date__month').annotate(sum=Sum('amount'))
+    categories = [
```
Can you add consts

LiorKGOW replied:
Sure 👍

ronyyosef commented on templates/house_view/mine_chart.html:
```diff
@@ -0,0 +1,28 @@
+<script type="text/javascript" src="{% static 'js/random_colors.js' %}" ></script>
+<script src="{%static 'js/pie_chart.js'%}" type="text/javascript"></script>
+<script src="{%static 'js/bar_chart.js'%}" type="text/javascript"></script>
+<script>
```
Can you please move the script to another file

LiorKGOW replied:
Can I reslove this conversation? @ronyyosef

ronyyosef replied:
Yes, thanks for explanation!

LiorKGOW commented:
> Nice work
> 
> The tests here coveting model tests, You had obviously updates the html templates, DYT, you should update client tests? 
> related to this? include 'house_view/mine_chart.html' include 'house_view/expenses_table.html'

Thanks @Yarboa 😃 

What did you mean by: "The tests here coveting model tests"
I haven't updated any models in this PR, so how could I be coveting them?
Maybe I miss understood your comment 😅 
I have added a test for the templates I have added: 'house_view/mine_chart.html', the other template is already checked in another PR that was already merged (#83 )👍,  I have just adjusted it's place ('house_view/expenses_table.html')
This PR changes the UI a bit, it shows the charts of the expenses of the year that is chosen in the choiceField below the left chart
which client tests needs to be changed?
Could you elaborate please? 😇

Yarboa commented:
> > Nice work
> > The tests here coveting model tests, You had obviously updates the html templates, DYT, you should update client tests?
> > related to this? include 'house_view/mine_chart.html' include 'house_view/expenses_table.html'
> 
> Thanks @Yarboa smiley
> 
> What did you mean by: "The tests here coveting model tests" I haven't updated any models in this PR, so how could I be coveting them? Maybe I miss understood your comment sweat_smile I have added a test for the templates I have added: 'house_view/mine_chart.html', the other template is already checked in another PR that was already merged (#83 )+1, I have just adjusted it's place ('house_view/expenses_table.html') This PR changes the UI a bit, it shows the charts of the expenses of the year that is chosen in the choiceField below the left chart which client tests needs to be changed? Could you elaborate please? innocent

Please add the links to tests covering this part

LiorKGOW commented:
> > > Nice work
> > > The tests here coveting model tests, You had obviously updates the html templates, DYT, you should update client tests?
> > > related to this? include 'house_view/mine_chart.html' include 'house_view/expenses_table.html'
> > 
> > 
> > Thanks @Yarboa smiley
> > What did you mean by: "The tests here coveting model tests" I haven't updated any models in this PR, so how could I be coveting them? Maybe I miss understood your comment sweat_smile I have added a test for the templates I have added: 'house_view/mine_chart.html', the other template is already checked in another PR that was already merged (#83 )+1, I have just adjusted it's place ('house_view/expenses_table.html') This PR changes the UI a bit, it shows the charts of the expenses of the year that is chosen in the choiceField below the left chart which client tests needs to be changed? Could you elaborate please? innocent
> 
> Please add the links to tests covering this part

tests/my_house_tests.py

LiorKGOW commented:
rebased and repushef after #139 was merged to main 👍

Omeramsc commented:
LGTM
@Yarboa please merge if you think it is ready. 
do note that's the last chance to merge anything

LiorKGOW commented:
@Yarboa I ahve updated the appropriate tests as you asked 👍 
please check it out 🙏 
https://github.com/redhat-beyond/Xpense/blob/5eb40498bb7a4eb5d51aaa28a41c31bdfa410f3e/tests/my_house_tests.py
